### PR TITLE
Store repair data under detached fields

### PR DESCRIPTION
**Pull request by jenn-le**
## Description

- allow field kind implementations to handle repair data
- for sequence field and optional field, translate changesets so that instead of generating delete deltas, they generate moves to a detached field instead
- add `TreeIndex` which associates a `ChangeAtomId` with a detached field key

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: experimental/dds/tree2/src/core/rebase/fence.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
 	"tags": ["rebase"],
 	"exports": ["index"],
-	"imports": ["tree", "util", "repair"]
+	"imports": ["tree", "util", "schema-stored"]
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/core/rebase/fence.json:4]: Why do we need the dependency on schema-stored?

File: experimental/dds/tree2/src/core/rebase/repairDataStore.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/core/rebase/repairDataStore.ts:1]: rebase seems like a surprising place for this (and repairDataStoreProvider). What was the motivation for moving it here (and for moving it at all)?

File: experimental/dds/tree2/src/core/rebase/fence.json (hunk 1 above)
Comment (jenn-le) [reply experimental/dds/tree2/src/core/rebase/fence.json:4]: FieldKey (used by the repair data store) has been moved to schema-stored. This is part of what I'll remove when we no longer need repair data store.

File: experimental/dds/tree2/src/core/tree/delta.ts [hunk 2]
```diff
@@ -224,6 +224,12 @@ export interface MoveOut<TTree = ProtoNode> extends HasModifications<TTree> {
 	 * The delta should carry exactly one `MoveIn` mark with the same move ID.
 	 */
 	readonly moveId: MoveId;
+
+	/**
+	 * When set, indicates that the delta was generated from a delete operation.
+	 * This means that the content being moved is repair data.
+	 */
+	readonly isDelete?: true;
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/core/tree/delta.ts:232]: Might as well embrace the new terminology:

```suggestion
	/**
	 * When set, indicates that the delta was generated from a remove operation.
	 * This means that the content being moved is repair data.
	 */
	readonly isRemoval?: true;
```

File: experimental/dds/tree2/src/core/rebase/index.ts [hunk 3]
```diff
@@ -32,3 +34,6 @@ export {
 	Violation,
 } from "./verifyChangeRebaser";
 export { findAncestor, findCommonAncestor, rebaseBranch, rebaseChange } from "./utils";
+// TODO: This is moved here temporarily to avoid a circular dependency. The RepairDataStore will be removed completely soon.
```
Comment (jenn-le) [inline experimental/dds/tree2/src/core/rebase/index.ts:37]: @yann-achard-MS This is the reasoning for the move. Let me know if you want me to do something else though.

File: experimental/dds/tree2/src/core/tree/visitDelta.ts [hunk 4]
```diff
@@ -175,6 +183,13 @@ function firstPass(delta: Delta.MarkList, visitor: DeltaVisitor, config: PassCon
 					markHasMoveOrDelete = true;
 					break;
 				case Delta.MarkType.MoveOut:
+					if (!config.allowRepairDataMoves && mark.isDelete) {
+						// Handled in the second pass
+						visitModify(index, mark, visitor, config);
+						index += mark.count;
+						break;
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/core/tree/visitDelta.ts:190]: Don't we need this as well?
```suggestion
						markHasMoveOrDelete = true;
						break;
```

File: experimental/dds/tree2/src/core/tree/visitDelta.ts [hunk 5]
```diff
@@ -240,6 +259,12 @@ function secondPass(delta: Delta.MarkList, visitor: DeltaVisitor, config: PassCo
 					}
 					break;
 				case Delta.MarkType.MoveIn: {
+					// If this move in is the result of a delete and should be treated as a normal delete,
+					// it is handled in the move out case instead.
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/core/tree/visitDelta.ts:263]: Nitpicking the language a bit:

```suggestion
					// If this move-in is the result of a removal and the visitor wishes to treat removals as deletions,
					// then there is no data to move in.
```

File: experimental/dds/tree2/src/feature-libraries/modular-schema/fieldChangeHandler.ts [hunk 6]
```diff
@@ -156,7 +161,11 @@ export interface FieldEditor<TChangeset> {
  * The `index` should be `undefined` iff the child node does not exist in the input context (e.g., an inserted node).
  * @alpha
  */
-export type ToDelta = (child: NodeChangeset) => Delta.Modify;
+export type ToDelta = (
+	child: NodeChangeset,
+	repairDataBuilder: RepairDataBuilder,
+	idAllocator: IdAllocator,
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/modular-schema/fieldChangeHandler.ts:167]: Could these parameters be automatically provided by the ModularChangeFamily implementation? This would save FieldKind implementation from having to pass them around.

File: experimental/dds/tree2/src/core/repair/repairDataBuilder.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import { ChangeAtomId } from "../rebase";
+import { FieldKey } from "../schema-stored";
+import { Delta } from "../tree";
+
+/**
+ * A function that associates a change that produces repair data with the detached
+ * field where the repair data should be stored.
+ * @alpha
+ */
+export type RepairDataHandler = (changeId: ChangeAtomId) => FieldKey;
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/core/repair/repairDataBuilder.ts:16]: How about renaming this to `RepairDataLocator`? "Locator" is not perfect either since it seems to imply that the repair data is already there, but it seems like a more specific name for the role it plays. I'm interested in other ideas you might have.

File: experimental/dds/tree2/src/core/repair/repairDataBuilder.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * TODO doc
+ * @alpha
+ */
+export interface RepairDataBuilder {
+	handler: RepairDataHandler;
+	/**
+	 * A map of detached fields to the delta that would move in the appropriate repair data.
+	 */
+	marks: Map<FieldKey, Delta.MarkList>;
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/core/repair/repairDataBuilder.ts:27]: Exposing the map to the code that adds content into it means they could theoretically remove content from it or condition their behavior based on what else is in the map. Consider tightening the contract with something like: `type RepairDataAccumulator = (key: FieldKey, delta: Delta.MarkList) => void`. You can still use a map behind the scenes.

File: experimental/dds/tree2/src/core/repair/repairDataBuilder.ts (hunk 7 above)
Comment (jenn-le) [reply experimental/dds/tree2/src/core/repair/repairDataBuilder.ts:16]: How about `RepairDataIndexer`?

File: experimental/dds/tree2/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 9]
```diff
@@ -712,43 +715,64 @@ export class ModularChangeFamily
...
 
-		return this.intoDeltaImpl(change.fieldChanges);
+		const repairDataBuilder = {
+			handler: repairDataHandler,
+			marks: repairDataMarks,
+		};
+		const idAllocator = idAllocatorFromMaxId(change.maxId);
+		const delta = this.intoDeltaImpl(change.fieldChanges, repairDataBuilder, idAllocator);
+
+		repairDataBuilder.marks.forEach((marks, field) => delta.set(field, marks));
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/modular-schema/modularChangeFamily.ts:738]: Instead of first collecting the marks in a separate map and then adding them all to the root delta, it make sense to directly add them to the root delta (see also the comment I added on RepairDataBuilder.marks).

File: experimental/dds/tree2/src/feature-libraries/modular-schema/fieldChangeHandler.ts (hunk 6 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/modular-schema/fieldChangeHandler.ts:167]: Actually it looks like your ModularChangeFamily changes already do what I had in mind, so I think these can just be removed because they're never used.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts [hunk 10]
```diff
@@ -15,17 +16,34 @@ import {
 	markIsTransient,
 } from "./utils";
 
-export type ToDelta<TNodeChange> = (child: TNodeChange) => Delta.Modify;
+export type ToDelta<TNodeChange> = (
+	child: TNodeChange,
+	repairDataBuilder: RepairDataBuilder,
+	idAllocator: IdAllocator,
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:22]: These can be removed as well.

File: experimental/dds/tree2/src/feature-libraries/deltaUtils.ts [hunk 11]
```diff
@@ -78,12 +78,20 @@ export function mapMark<TIn, TOut>(
 			};
 		}
 		case Delta.MarkType.MoveOut: {
-			return {
-				type: Delta.MarkType.MoveOut,
-				count: mark.count,
-				moveId: mark.moveId,
-				...mapModifications(mark, func),
-			};
+			return mark.isDelete
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/deltaUtils.ts:81]: I recommend creating a local witout the `isDelete` populated first, and then doing:
```
if (mark.isDelete) {
	delta.isDelete = true;
}
```
It avoids the code duplication.

File: experimental/dds/tree2/src/feature-libraries/treeIndex.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+	private readonly detachedFields = new Map<ChangeAtomId, FieldKey>();
+
+	public constructor(private readonly name: string) {}
+
+	/**
+	 * Returns a field key for the given change atom ID and ID. Should be unique for
+	 * this index.
+	 */
+	public getFieldKey(changeAtomId: ChangeAtomId, id: string): FieldKey {
+		assert(!this.detachedFields.has(changeAtomId), "Detached field already exists");
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/treeIndex.ts:23]: I think this assert may need to be removed for the sake of the restore (old name: "revive") codepath.

File: experimental/dds/tree2/src/shared-tree-core/sharedTreeCore.ts [hunk 13]
```diff
@@ -191,7 +192,10 @@ export class SharedTreeCore<TEditor extends ChangeFamilyEditor, TChange> extends
 			}
 
 			if (change !== undefined) {
-				this.changeEvents.emit("newLocalState", this.changeFamily.intoDelta(change));
+				this.changeEvents.emit(
+					"newLocalState",
+					this.changeFamily.intoDelta(change, () => EmptyKey),
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree-core/sharedTreeCore.ts:197]: Isn't this a problem? If that delta is used on a forest it will not create the correct repair data.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts [hunk 14]
```diff
@@ -74,9 +94,24 @@ function cellDeltaFromMark<TNodeChange>(
...
 			case "Delete": {
+				const detachedField = repairDataBuilder.handler({
+					revision: mark.revision,
+					localId: mark.id,
+				});
+				const moveId = brandOpaque<Delta.MoveId>(idAllocator());
+				repairDataBuilder.marks.set(detachedField, [
+					{
+						type: Delta.MarkType.MoveIn,
+						count: mark.count,
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:105]: Unfortunately, we need to create `mark.count` move-out/move-in pairs of count 1 instead of creating a single pair of count `mark.count`.
This is because addressing the repair data will not work well otherwise: if the nth node is revived then all the nodes after that would end up at a shifted index.
We'll add support for batching later, but for now a loop will do.

File: experimental/dds/tree2/src/core/rebase/types.ts [hunk 15]
```diff
@@ -24,6 +24,33 @@ export const SessionIdSchema = brandedStringType<SessionId>();
 export type RevisionTag = StableId;
 export const RevisionTagSchema = brandedStringType<StableId>();
 
+/**
+ * An ID which is unique within a revision of a changeset.
+ * A changeset which is a composition of multiple revisions may contain duplicate `ChangesetLocalId`s,
+ * but they are unique when qualified by the revision of the change they are used in.
+ * @alpha
+ */
+export type ChangesetLocalId = Brand<number, "ChangesetLocalId">;
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/core/rebase/types.ts:33]: Moving this would have been better done as a separate PR to reduce the noise and avoid the merge conflicts. Up to you if you want to keep it bundled or not.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts [hunk 16]
```diff
@@ -74,9 +94,24 @@ function cellDeltaFromMark<TNodeChange>(
 				return mark.count;
 			}
 			case "Delete": {
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:96]: The Revive case (on which GH doesn't let me put a comment) also needs to be updated to produce the inverse move.

File: experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/optionalField.spec.ts [hunk 17]
```diff
@@ -53,15 +57,25 @@ const failCrossFieldManager: CrossFieldManager = {
...
+const deltaFromChild1 = (
+	child: NodeChangeset,
+	{ marks, handler }: RepairDataBuilder,
+	idAllocator: IdAllocator,
+): Delta.Modify => {
 	assert.deepEqual(child, nodeChange1);
+
+	const moveId = brandOpaque<Delta.MoveId>(idAllocator());
+	marks.set(handler({ revision: change1Revision, localId: brand(0) }), [
+		{ type: Delta.MarkType.MoveIn, count: 1, moveId },
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/optionalField.spec.ts:69]: Don't forget the isDelete/isRemoval flag here and for the MoveOut, and for the deltaFromChild2 implementation.

File: experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/optionalField.spec.ts [hunk 18]
```diff
@@ -337,7 +363,7 @@ describe("optionalField", () => {
 						[
 							fooKey,
 							[
-								{ type: Delta.MarkType.Delete, count: 1 },
+								{ type: Delta.MarkType.MoveOut, count: 1, moveId: brand(0) },
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/optionalField.spec.ts:366]: I don't understand why this test expects there to be a deletion in the delta even though the field starts empty. This seems like a bug.
I'll look into it and make a separate PR to fix the code and the test expectations.

Comment (yann-achard-MS) [reply experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/optionalField.spec.ts:366]: Ah, my bad, this is for the nested change.

File: experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/optionalField.spec.ts [hunk 19]
```diff
@@ -348,17 +374,108 @@ describe("optionalField", () => {
...
 		it("can be converted to a delta when restoring content", () => {
 			const expected: Delta.MarkList = [
-				{ type: Delta.MarkType.Delete, count: 1 },
+				{ type: Delta.MarkType.MoveOut, count: 1, moveId: brand(0), isDelete: true },
 				{ type: Delta.MarkType.Insert, content: [testTreeCursor("tree1")] },
 			];
 
-			const actual = optionalFieldIntoDelta(revertChange2, deltaFromChild1);
+			const expectedRepairData = new Map<FieldKey, Delta.MarkList>([
+				[
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/optionalField.spec.ts:457]: There should also be a move-out from the repair data.

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceFieldToDelta.spec.ts [hunk 20]
```diff
@@ -120,7 +139,13 @@ describe("SequenceField - toDelta", () => {
...
+		const { repairDataBuilder } = makeRepairDataBuilder();
+		const actual = SF.sequenceFieldToDelta(
+			changeset,
+			deltaFromChild,
+			repairDataBuilder,
+			idAllocatorFromMaxId(brand(3)),
+		);
 		const expected: Delta.MarkList = [
 			{
 				type: Delta.MarkType.Insert,
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceFieldToDelta.spec.ts:151]: This should become a MoveIn from a repair data detached field.

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceFieldToDelta.spec.ts [hunk 21]
```diff
@@ -129,17 +154,29 @@ describe("SequenceField - toDelta", () => {
 			},
 		];
 		assertMarkListEqual(actual, expected);
+		assert.deepEqual(repairDataBuilder.marks, new Map([]));
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceFieldToDelta.spec.ts:157]: There should be a MoveOut in here.

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceFieldToDelta.spec.ts [hunk 22]
```diff
@@ -129,17 +154,29 @@ describe("SequenceField - toDelta", () => {
...
 	});
 
-	it("delete", () => {
+	it("delete produces repair data", () => {
 		const changeset = Change.delete(0, 10);
-		const mark: Delta.Delete = {
-			type: Delta.MarkType.Delete,
+		const mark: Delta.MoveOut = {
+			type: Delta.MarkType.MoveOut,
+			count: 10,
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceFieldToDelta.spec.ts:164]: I recommend lowering the count to 2 so you don't have to type out a ton of raw data in the test.

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceFieldToDelta.spec.ts [hunk 23]
```diff
@@ -237,10 +286,14 @@ describe("SequenceField - toDelta", () => {
 	});
 
 	it("modify and delete => delete", () => {
-		const changeset = [Mark.delete(1, brand(0), { changes: childChange1 })];
-		const mark: Delta.Delete = {
-			type: Delta.MarkType.Delete,
+		const changeset: TestChangeset = [
+			{ type: "Delete", id: brand(0), count: 1, changes: childChange1 },
+		];
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceFieldToDelta.spec.ts:291]: Try to use the Mark.delete instead of making changeset literals by hand. This may have been from a bad merge.

File: experimental/dds/tree2/src/shared-tree-core/sharedTreeCore.ts (hunk 13 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/shared-tree-core/sharedTreeCore.ts:197]: Shouldn't this use the localBranch's RepairDataHandler?

File: experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/optionalField.spec.ts (hunk 19 above)
Comment (jenn-le) [reply experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/optionalField.spec.ts:457]: The move out is part of `expected`. I check them separately because the move in gets added to the `repairDataMarks` map via the accumulator.

File: experimental/dds/tree2/src/core/rebase/fence.json (hunk 1 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/core/rebase/fence.json:4]: Ok. Please update the dependency diagram in the README if need be.

File: experimental/dds/tree2/src/core/rebase/index.ts (hunk 3 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/core/rebase/index.ts:37]: What is the circular chain in this case?

File: experimental/dds/tree2/src/core/rebase/repairDataStore.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/core/rebase/repairDataStore.ts:1]: Resolving. See conversation in index.ts

File: experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/optionalField.spec.ts (hunk 19 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/optionalField.spec.ts:457]: No, I was referring to the MoveOut that would be necessary in `expectedRepairData` for the sake of restoring the content. This falls under the "do we support revive as part of this PR?" bucket.

File: experimental/dds/tree2/src/feature-libraries/modular-schema/fieldChangeHandler.ts (hunk 6 above)
Comment (jenn-le) [reply experimental/dds/tree2/src/feature-libraries/modular-schema/fieldChangeHandler.ts:167]: They're meant to be used by the child field kind implementations. See the deltaFromChilds in `optionalFieldKind.spec.ts`. I did it this way cause I was thinking that any field kind implementation could need to generate repair deltas so we'd need to pass this info when generating child deltas. Is this not necessary or is there a way for the ToDeltas to access this from the ModularChangeFamily?

Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/modular-schema/fieldChangeHandler.ts:167]: You're indeed correct that any field kind implementation might need access to those, but this API is not about passing those to a field kind implementation, it is about passing those back to ModularChangeFamily (which provided them in the first place). If you look at the code you currently have in modularChangeFamily.ts, `intoDeltaImpl` doesn't use those parameters:
```
(childChange /* parms 2 and 3 are ignored */): Delta.Modify => this.deltaFromNodeChange(childChange, repairDataBuilder, idAllocator)
```

You are correct that removing those two parameters from `ToDelta` means that `deltaFromChild1` and `deltaFromChild2` need to change. I think you can make a helper function that take the RepairDataBuilder and the IdAllocator, and returns a `deltaFromChild` function that is seeded with the appropriate RepairDataBuilder and IdAllocator. It may even make sense for that helper function to build the RepairDataBuilder in the first place (and possibly the IdAllocator too, I'm not sure) and retrns them too.

File: experimental/dds/tree2/src/core/rebase/index.ts (hunk 3 above)
Comment (jenn-le) [reply experimental/dds/tree2/src/core/rebase/index.ts:37]: Repair needs rebase now because that's where we're exposing change atom id (also changed this in the README chart). I figured this would be reasonable since we'll be removing the repair stuff that rebase needs.

Comment (yann-achard-MS) [reply experimental/dds/tree2/src/core/rebase/index.ts:37]: I just wasn't sure whether there were more links in the chain that forms the cycles, which made it hard to evaluate things like "Should some of these link be merged into a single directory?". After looking it up I can tell that's not the case. I agree this is fine. Marking this as resolved.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts (hunk 16 above)
Comment (jenn-le) [reply experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:101]: This was discussed offline and will be done in a separate PR.

File: experimental/dds/tree2/src/shared-tree-core/sharedTreeCore.ts (hunk 13 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/shared-tree-core/sharedTreeCore.ts:197]: This will be removed right after this PR, so that's OK.

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceFieldToDelta.spec.ts (hunk 20 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceFieldToDelta.spec.ts:157]: Will be tackled later.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts [hunk 24]
```diff
@@ -74,10 +99,29 @@ function cellDeltaFromMark<TNodeChange>(
...
-				return {
-					type: Delta.MarkType.Delete,
-					count: mark.count,
-				};
+				const detachedField = repairDataBuilder.handler({
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:102]: Each removed node needs to go in a separate detached field that corresponds to the ChangeAtomId for that node. So you'll need to move that code to be inside the loop and do `localId: mark.id + i` (modulo whatever the type system wants to do in order to make it type-check).

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts [hunk 25]
```diff
@@ -74,10 +99,29 @@ function cellDeltaFromMark<TNodeChange>(
...
+						isRemoval: true,
+					});
+					moveInMarks.push({
+						type: Delta.MarkType.MoveIn,
+						count: 1,
+						moveId,
+						isRemoval: true,
+					});
+				}
+				repairDataBuilder.accumulator(detachedField, moveInMarks);
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:123]: That will need to be pulled inside the loop as well.

File: experimental/dds/tree2/src/feature-libraries/treeIndex.ts [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+ * ID used to create a detached field key for a removed subtree.
+ * @alpha
+ */
+export type RemovedTreeId = Brand<string, "tree.RemovedTreeId">;
+
+/**
+ * The tree index records detached field ids and associates them with a change atom ID.
+ */
+export class TreeIndex {
+	private readonly detachedFields = new SizedNestedMap<
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/treeIndex.ts:19]: Are you expecting that we'll want to query the size of the map?
